apis/keda/v1alpha1/scaledobject_webhook.go

```diff
@@ -0,0 +1,197 @@
+// +kubebuilder:rbac:groups="",namespace=keda,resources=secrets,verbs=get;list;watch;create;update;patch;delete
+
+var _ webhook.Validator = &ScaledObject{}
+
+// ValidateCreate implements webhook.Validator so a webhook will be registered for the type
+func (so *ScaledObject) ValidateCreate() error {
+	val, _ := json.MarshalIndent(so, "", "  ")
+	scaledobjectlog.V(1).Info(fmt.Sprintf("validating scaledobject creation for %s", string(val)))
+	return validateWorkload(so, "create")
+}
+
+func (so *ScaledObject) ValidateUpdate(old runtime.Object) error {
+	val, _ := json.MarshalIndent(so, "", "  ")
+	scaledobjectlog.V(1).Info(fmt.Sprintf("validating scaledobject update for %s", string(val)))
+	return validateWorkload(so, "update")
+}
+
+func validateWorkload(so *ScaledObject, action string) error {
+	prommetrics.RecordScaledObjectValidatingTotal(so.Namespace, action)
+	err := verifyScaledObjects(so, action)
+	if err != nil {
+		return err
+	}
+	return verifyHpas(so, action)
+}
+
+func verifyHpas(incomingSo *ScaledObject, action string) error {
+	hpaList := &autoscalingv2.HorizontalPodAutoscalerList{}
+	opt := &client.ListOptions{
+		Namespace: incomingSo.Namespace,
+	}
+	err := kc.List(context.Background(), hpaList, opt)
+	if err != nil {
+		return err
+	}
+
+	for _, hpa := range hpaList.Items {
+		val, _ := json.MarshalIndent(hpa, "", "  ")
+		scaledobjectlog.V(1).Info(fmt.Sprintf("checking hpa %s: %v", hpa.Name, string(val)))
+		hpaTarget := hpa.Spec.ScaleTargetRef
+		incomingSoTarget := incomingSo.Spec.ScaleTargetRef
+
+		// prepare default values
+		hpatargetAPI := defaultAPI
+		if hpaTarget.APIVersion != "" {
+			hpatargetAPI = hpaTarget.APIVersion
+		}
+		hpaTargetKind := defaultKind
+		if hpaTarget.Kind != "" {
+			hpaTargetKind = hpaTarget.Kind
+		}
+		incomingSotargetAPI := defaultAPI
+		if incomingSoTarget.APIVersion != "" {
+			incomingSotargetAPI = incomingSoTarget.APIVersion
+		}
+		incomingSoTargetKind := defaultKind
+		if incomingSoTarget.Kind != "" {
+			incomingSoTargetKind = incomingSoTarget.Kind
+		}
+
+		if hpatargetAPI == incomingSotargetAPI &&
+			hpaTargetKind == incomingSoTargetKind &&
+			hpaTarget.Name == incomingSoTarget.Name {
+			owned := false
+			ownerName := ""
+			for _, owner := range hpa.OwnerReferences {
+				if owner.Kind == incomingSo.Kind {
+					ownerName = owner.Name
+					if owner.Name == incomingSo.Name {
+						owned = true
+					}
+				}
+			}
+
+			if !owned {
+				var err error
+				if len(hpa.OwnerReferences) == 0 {
+					err = fmt.Errorf("the workload '%s' of type '%s/%s' is already managed by the hpa '%s'", incomingSoTarget.Name, incomingSotargetAPI, incomingSoTargetKind, hpa.Name)
+				} else {
+					err = fmt.Errorf("the workload '%s' of type '%s/%s' is already managed by the ScaledObject '%s'", incomingSoTarget.Name, incomingSotargetAPI, incomingSoTargetKind, ownerName)
```

Comment:
What if the HPA is owned by some other resource (not just ScaledObject)? This should message should be imho generic and api/kind/name should be parametrized.

Reply:
This code is old and I forgot removing it. The first approach was to check all in the same function, but now we are checking 1st the SOs conflicts and then the HPA conflicts, so this else should never happen. If the code reach here, only the HPA case can happen